### ISSUE TITLE
Propagate or generate transaction ID

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#1916]
In the case of receiving a transaction ID (in the incoming HTTP request in a given header), it has to be used (and propagated to any outgoing HTTP request derived from such incoming HTTP request) instead of auto-generating one. 

If no transaction ID is provided in the HTTP request then CB has to auto-generate one and propagate it to any outgoing HTTP request derived from that transaction (forwarding and notifications).

Details about which particular header to use for transaction ID are under discussion in the moment of writing this.

**Comment by kzangeli** [comment]
I imagine that we will use the same transaction ID for forwarding messages also.

What about notifications?
Will they carry the transaction ID of the update that provoked the notification?

**Comment by fgalan** [comment]
> Will they carry the transaction ID of the update that provoked the notification?

That's correct, in the same way the for example carry the same X-Auth-Token header that the one in the updated provoking the notification.

**Comment by fgalan** [comment]
At the end, the name of the header is going to be:

```
Fiware-Transaction
```

**Comment by fgalan** [comment]
Currently we are using 

```
<CB start timestamp (seconds)>-<CB start timestamp (ms)>-<transaction number from startup>
```

eg: 1459515461-844-00000000001

However, maybe at the end we will use random UUIDv4 (https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_.28random.29), which use the following format:

We should use existing libraries for generation, such libuuid (http://linux.die.net/man/3/libuuid) which seems to be available at CentOS after `yum search libuuid`.

See also: http://stackoverflow.com/questions/16884288/which-uuid-library-to-use

**Comment by fgalan** [comment]
However, note that the internal transaction counter (either auto-generated or re-transmited) will be still in use in order to fill the information in `SUMMARY` traces.

**Comment by fgalan** [comment]
Additional considerations:

* The field to map Fiware-Transaction is corr=
* The old transactionID= (or trans=... I don't remember exact name right now) would be "N/A" from now on. In the future it could be use for internal ID corresponding to the (global) transaction, but that is another history...

**Comment by fgalan** [comment]
Final name for the header: `Fiware-Correlator`.

**Comment by fgalan** [comment]
After discussion, the correlator (either auto-generated or received in the Fiware-Correlator header in the request) has to be included in the response.

**Comment by fgalan** [comment]
Implemented in PR #2025. Asigned to QATestPending issue until corresponding functional test gets implemented.

**Comment by fgalan** [comment]
Issues in milestone QATestPendingNotE2ECovered are going to be closed (it is not realistic to think we would have people working on behave testing at the present moment). However, issue is labeled with QATestPending label in the case things change and we need to recover it.
